File: clients/cpp/model/WorkerGroup.cpp
#include "WorkerGroup.hpp"

namespace model {

WorkerGroup::WorkerGroup(int playerIndex, int number) : playerIndex(playerIndex), number(number) { }

// Read WorkerGroup from input stream
WorkerGroup WorkerGroup::readFrom(InputStream& stream) {
    int playerIndex = stream.readInt();
    int number = stream.readInt();
    return WorkerGroup(playerIndex, number);
}

// Write WorkerGroup to output stream
void WorkerGroup::writeTo(OutputStream& stream) const {
    stream.write(playerIndex);
    stream.write(number);
}

// Get string representation of WorkerGroup
std::string WorkerGroup::toString() const {
    std::stringstream ss;
    ss << "WorkerGroup { ";
    ss << "playerIndex: ";
    ss << playerIndex;
    ss << ", ";
    ss << "number: ";
    ss << number;
    ss << " }";
    return ss.str();
}

bool WorkerGroup::operator ==(const WorkerGroup& other) const {
    return playerIndex == other.playerIndex && number == other.number;
}

}

size_t std::hash<model::WorkerGroup>::operator ()(const model::WorkerGroup& value) const {
    size_t result = 0;
    result ^= std::hash<int>{}(value.playerIndex) + 0x9e3779b9 + (result << 6) + (result >> 2);
    result ^= std::hash<int>{}(value.number) + 0x9e3779b9 + (result << 6) + (result >> 2);
    return result;
}

File: clients/rust/src/model/worker_group.rs
use super::*;

/// TODO - Document
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct WorkerGroup {
    /// TODO - Document
    pub player_index: i32,
    /// TODO - Document
    pub number: i32,
}

impl trans::Trans for WorkerGroup {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.player_index.write_to(writer)?;
        self.number.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let player_index: i32 = trans::Trans::read_from(reader)?;
        let number: i32 = trans::Trans::read_from(reader)?;
        Ok(Self {
            player_index,
            number,
        })
    }
}